### PR TITLE
docs: add Go to the documentation

### PR DESCRIPTION
**Pull request by mdelapenya**
# Description

This PR is adding examples on how to organise the Go code using [godog](https://github.com/cucumber/godog), the official implementation of Cucumber in Go.

It only applies to the "Step Organisation" section of the docs, as I wanted to first receive feedback about it. I can elaborate a follow-up PR including more updates.

# Motivation & context

I was following the tutorial, trying to reuse some parts for an internal tutorial and missed the Go version, as we are using it internally.

## Type of change

- Docs: add documentation for Go

_Remember to add an entry to the relevant section of CHANGELOG.md as part of this pull request._

## Note to other contributors

_If your change may impact future contributors, explain it here, and remember to update README.md and CONTRIBUTING.md accordingly._

## Update required of cucumber.io/docs

_If the [Cucumber documentation](https://cucumber.io/docs/) will require an update,
submit an issue or ideally a pull request to [cucumber/docs](https://github.com/cucumber/docs/) and
reference it here._

_e.g. "Ref: cucumber/docs/pull/#99"_

# Checklist:

- [ ] I have read the [**CONTRIBUTING**](../CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] I have added tests to cover my changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../CHANGELOG.md), linking to this pull request.

### CODE REVIEW

File: content/docs/gherkin/step-organization.md [hunk 1]
```diff
@@ -15,16 +16,16 @@ This will make your project more logical and easier to maintain.
...
 Be aware that, regardless of the directory structure employed, Cucumber effectively flattens the `features/` directory tree when running tests.
-This means that anything ending in {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}
+This means that anything ending in {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}{{% text "go" %}}`.go`{{% /text %}}
 inside the directory in which Cucumber is run is treated as a step definition. In the same directory, Cucumber will search for a `Feature` corresponding to that step definition.
-This is either the default case or the location specified with the {{% text "java,kotlin" %}}relevant{{% /text %}}{{% text "javascript" %}}relevant{{% /text %}}{{% text "ruby" %}}`-r`{{% /text %}} option.
+This is either the default case or the location specified with the {{% text "java,kotlin" %}}relevant{{% /text %}}{{% text "javascript" %}}relevant{{% /text %}}{{% text "ruby" %}}`-r`{{% /text %}}{{% text "go" %}}relevant{{% /go %}} option.
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
This is either the default case or the location specified with the {{% text "java,kotlin" %}}relevant{{% /text %}}{{% text "javascript" %}}relevant{{% /text %}}{{% text "ruby" %}}`-r`{{% /text %}}{{% text "go" %}}relevant{{% /text %}} option.
```

File: content/docs/gherkin/step-organization.md [hunk 2]
```diff
@@ -6,6 +6,7 @@ polyglot:
  - javascript
  - ruby
  - kotlin
+ - go
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
 - golang
```

File: content/docs/gherkin/step-organization.md [hunk 3]
```diff
@@ -15,16 +16,16 @@ This will make your project more logical and easier to maintain.
 
 # How Cucumber finds your features and step definitions
 Be aware that, regardless of the directory structure employed, Cucumber effectively flattens the `features/` directory tree when running tests.
-This means that anything ending in {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}
+This means that anything ending in {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}{{% text "go" %}}`.go`{{% /text %}}
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
This means that anything ending in {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}{{% text "golang" %}}`.go`{{% /text %}}
```

File: content/docs/gherkin/step-organization.md [hunk 4]
```diff
@@ -15,16 +16,16 @@ This will make your project more logical and easier to maintain.
...
 inside the directory in which Cucumber is run is treated as a step definition. In the same directory, Cucumber will search for a `Feature` corresponding to that step definition.
-This is either the default case or the location specified with the {{% text "java,kotlin" %}}relevant{{% /text %}}{{% text "javascript" %}}relevant{{% /text %}}{{% text "ruby" %}}`-r`{{% /text %}} option.
+This is either the default case or the location specified with the {{% text "java,kotlin" %}}relevant{{% /text %}}{{% text "javascript" %}}relevant{{% /text %}}{{% text "ruby" %}}`-r`{{% /text %}}{{% text "go" %}}relevant{{% /text %}} option.
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
This is either the default case or the location specified with the {{% text "java,kotlin" %}}relevant{{% /text %}}{{% text "javascript" %}}relevant{{% /text %}}{{% text "ruby" %}}`-r`{{% /text %}}{{% text "golang" %}}relevant{{% /text %}} option.
```

File: content/docs/gherkin/step-organization.md [hunk 5]
```diff
@@ -15,16 +16,16 @@ This will make your project more logical and easier to maintain.
...
 
 # Grouping step definitions
 Technically it doesn't matter how you name your step definition files, or which step definitions you put in a file.
 You *could* have one giant file containing all your step definitions. However, as the project grows, the file can become messy and hard to maintain.
-Instead, we recommend creating a separate{{% text "ruby" %}} `*_steps.rb`{{% /text %}}{{% text "java" %}} `StepDefinitions.java`{{% /text %}}{{% text "kotlin" %}} `StepDefinitions.kt`{{% /text %}}{{% text "javascript" %}} `*_steps.js`{{% /text %}} file for each domain concept.
+Instead, we recommend creating a separate{{% text "ruby" %}} `*_steps.rb`{{% /text %}}{{% text "java" %}} `StepDefinitions.java`{{% /text %}}{{% text "kotlin" %}} `StepDefinitions.kt`{{% /text %}}{{% text "javascript" %}} `*_steps.js`{{% /text %}}{{% text "go" %}} `*_steps.go`{{% /text %}} file for each domain concept.
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
Instead, we recommend creating a separate{{% text "ruby" %}} `*_steps.rb`{{% /text %}}{{% text "java" %}} `StepDefinitions.java`{{% /text %}}{{% text "kotlin" %}} `StepDefinitions.kt`{{% /text %}}{{% text "javascript" %}} `*_steps.js`{{% /text %}}{{% text "golang" %}} `*_steps.go`{{% /text %}} file for each domain concept.
```

File: content/docs/gherkin/step-organization.md [hunk 6]
```diff
@@ -15,16 +16,16 @@ This will make your project more logical and easier to maintain.
...
 
-A good rule of thumb is to have one file for each major {{% text "ruby" %}}model/database table.{{% /text %}}{{% text "java,kotlin" %}}domain object.{{% /text %}}{{% text "javascript" %}}domain object.{{% /text %}}
+A good rule of thumb is to have one file for each major {{% text "ruby" %}}model/database table.{{% /text %}}{{% text "java,kotlin" %}}domain object.{{% /text %}}{{% text "javascript" %}}domain object.{{% /text %}}{{% text "go" %}}domain object.{{% /text %}}
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
A good rule of thumb is to have one file for each major {{% text "ruby" %}}model/database table.{{% /text %}}{{% text "java,kotlin" %}}domain object.{{% /text %}}{{% text "javascript" %}}domain object.{{% /text %}}{{% text "golang" %}}domain object.{{% /text %}}
```

File: content/docs/gherkin/step-organization.md [hunk 7]
```diff
@@ -51,8 +52,14 @@ For example, in a Curriculum Vitae application, we might have:
...
 {{% /block %}}
+{{% block "go" %}}
+- `employee_steps.go`
+- `education_steps.go`
+- `experience_steps.go`
+- `authentication_steps.go`
+{{% /block %}}
 
-The first three files would define all the `Given`, `When`, and `Then` step definitions related to creating, reading, updating, and deleting the various {{% text "ruby" %}}models.{{% /text %}}{{% text "java" %}}types of objects.{{% /text %}}{{% text "javascript" %}}types of objects.{{% /text %}}
+The first three files would define all the `Given`, `When`, and `Then` step definitions related to creating, reading, updating, and deleting the various {{% text "ruby" %}}models.{{% /text %}}{{% text "java,go" %}}types of objects.{{% /text %}}{{% text "javascript" %}}types of objects.{{% /text %}}
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
The first three files would define all the `Given`, `When`, and `Then` step definitions related to creating, reading, updating, and deleting the various {{% text "ruby" %}}models.{{% /text %}}{{% text "java,golang" %}}types of objects.{{% /text %}}{{% text "javascript" %}}types of objects.{{% /text %}}
```

File: content/docs/gherkin/step-organization.md [hunk 8]
```diff
@@ -51,8 +52,14 @@ For example, in a Curriculum Vitae application, we might have:
 - `experience_steps.js`
 - `authentication_steps.js`
 {{% /block %}}
+{{% block "go" %}}
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
{{% block "golang" %}}
```

File: content/docs/gherkin/step-organization.md [hunk 9]
```diff
@@ -124,7 +139,7 @@ This helps in a number of ways:
 
 You can handle other behaviours, like *validating a web page, clicking a button, etc.*, the same way.
 
-{{% text "java,kotlin" %}}We suggest taking a look at the [Factory Design Pattern] (https://refactoring.guru/design-patterns/factory-method) as well.{{% /text %}}
+{{% text "java,kotlin,go" %}}We suggest taking a look at the [Factory Design Pattern] (https://refactoring.guru/design-patterns/factory-method) as well.{{% /text %}}
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
{{% text "java,kotlin,golang" %}}We suggest taking a look at the [Factory Design Pattern] (https://refactoring.guru/design-patterns/factory-method) as well.{{% /text %}}
```

File: content/docs/gherkin/step-organization.md [hunk 10]
```diff
@@ -134,6 +149,6 @@ On the other hand, do not lose sight that every step called as such in a step de
 [Gherkin](/docs/gherkin/) and therefore must conform to the same syntax as used in feature files.
 
 In fact, it is recommended to refactor step definitions into helper methods for greater modularity and reuse.
-The method can reside in the same {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}} file as the step definition.
+The method can reside in the same {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}{{% text "go" %}}`.go`{{% /text %}} file as the step definition.
```
Comment (mdelapenya) [inline content/docs/gherkin/step-organization.md]:
```suggestion
The method can reside in the same {{% text "java" %}}`.java`{{% /text %}}{{% text "kotlin" %}}`.kt`{{% /text %}}{{% text "javascript" %}}`.js`{{% /text %}}{{% text "ruby" %}}`.rb`{{% /text %}}{{% text "golang" %}}`.go`{{% /text %}} file as the step definition.
```

**Comment by aurelien-reeves** [comment]
Great idea, great PR! 

I agree to merge it as-is and to follow-up with other PR for adding go-lang to other pages.
But I just would like to wait for other contributors opinion on that :p

**Review by aurelien-reeves — APPROVED** [review]

**Review by aurelien-reeves — CHANGES_REQUESTED** [review]
~~The golang logo is missing actually 😅~~

**Comment by aurelien-reeves** [comment]
> 



> The golang logo is missing actually 😅

Oh, actually it is not missing, sorry for that 😓

**Review by aurelien-reeves — APPROVED** [review]

**Comment by aslakhellesoy** [comment]
Hi @mdelapenya,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Pull request merged**